File: apps/python-api/app/repository.py
from __future__ import annotations

import json
from typing import Any

from psycopg import Connection
from psycopg.errors import UniqueViolation
from psycopg_pool import ConnectionPool

from . import queries
from .errors import ApiError
from .serializers import customer_from_row, order_from_rows, product_from_row


def _fetch_customer(conn: Connection, customer_id: int) -> dict[str, Any] | None:
    row = conn.execute(queries.GET_CUSTOMER, (customer_id,)).fetchone()
    if row is None:
        return None
    return customer_from_row(row)


def get_customer(pool: ConnectionPool, customer_id: int) -> dict[str, Any]:
    with pool.connection() as conn:
        customer = _fetch_customer(conn, customer_id)
    if customer is None:
        raise ApiError(404, "NOT_FOUND", "Customer not found")
    return customer


def list_customers(pool: ConnectionPool, page: int, page_size: int) -> dict[str, Any]:
    offset = (page - 1) * page_size
    with pool.connection() as conn:
        total = conn.execute(queries.COUNT_CUSTOMERS).fetchone()["total"]
        rows = conn.execute(queries.LIST_CUSTOMERS, (page_size, offset)).fetchall()
    return {
        "page": page,
        "pageSize": page_size,
        "total": total,
        "items": [customer_from_row(row) for row in rows],
    }


def create_customer(pool: ConnectionPool, payload: dict[str, Any]) -> dict[str, Any]:
    try:
        with pool.connection() as conn:
            with conn.transaction():
                customer_id = conn.execute(
                    queries.INSERT_CUSTOMER,
                    (
                        payload["fullName"],
                        payload["email"],
                        payload["documentNumber"],
                        payload.get("phone"),
                    ),
                ).fetchone()["id"]
                address = payload["address"]
                conn.execute(
                    queries.INSERT_ADDRESS,
                    (
                        customer_id,
                        address["label"],
                        address["street"],
                        address["number"],
                        address.get("complement"),
                        address["district"],
                        address["city"],
                        address["state"],
                        address["postalCode"],
                        address["isDefault"],
                    ),
                )
                conn.execute(
                    queries.INSERT_AUDIT_LOG,
                    ("customer", customer_id, "create_customer", json.dumps(payload, ensure_ascii=False)),
                )
                customer = _fetch_customer(conn, customer_id)
    except UniqueViolation as exc:
        raise ApiError(409, "CONFLICT", "Customer email or document already exists") from exc
    if customer is None:
        raise ApiError(500, "DATABASE_ERROR", "Customer was created but could not be loaded")
    return customer


def update_customer(pool: ConnectionPool, customer_id: int, payload: dict[str, Any]) -> dict[str, Any]:
    with pool.connection() as conn:
        with conn.transaction():
            updated = conn.execute(
                queries.UPDATE_CUSTOMER,
                (
                    payload["fullName"],
                    payload.get("phone"),
                    payload["status"],
                    customer_id,
                ),
            ).fetchone()
            if updated is None:
                raise ApiError(404, "NOT_FOUND", "Customer not found")

            address = payload["address"]
            changed_address = conn.execute(
                queries.UPDATE_DEFAULT_ADDRESS,
                (
                    address["label"],
                    address["street"],
                    address["number"],
                    address.get("complement"),
                    address["district"],
                    address["city"],
                    address["state"],
                    address["postalCode"],
                    address["isDefault"],
                    customer_id,
                ),
            ).fetchone()
            if changed_address is None:
                conn.execute(
                    queries.INSERT_ADDRESS,
                    (
                        customer_id,
                        address["label"],
                        address["street"],
                        address["number"],
                        address.get("complement"),
                        address["district"],
                        address["city"],
                        address["state"],
                        address["postalCode"],
                        address["isDefault"],
                    ),
                )

            conn.execute(
                queries.INSERT_AUDIT_LOG,
                ("customer", customer_id, "update_customer", json.dumps(payload, ensure_ascii=False)),
            )
            customer = _fetch_customer(conn, customer_id)
    if customer is None:
        raise ApiError(500, "DATABASE_ERROR", "Customer was updated but could not be loaded")
    return customer


def list_products(pool: ConnectionPool, category_id: int) -> dict[str, Any]:
    with pool.connection() as conn:
        category = conn.execute(queries.CATEGORY_EXISTS, (category_id,)).fetchone()
        if category is None:
            raise ApiError(404, "NOT_FOUND", "Category not found")
        rows = conn.execute(queries.LIST_PRODUCTS_BY_CATEGORY, (category_id,)).fetchall()
    return {
        "categoryId": category_id,
        "items": [product_from_row(row) for row in rows],
    }


def _fetch_order(conn: Connection, order_id: int) -> dict[str, Any] | None:
    rows = conn.execute(queries.GET_ORDER, (order_id,)).fetchall()
    return order_from_rows(rows)


def get_order(pool: ConnectionPool, order_id: int) -> dict[str, Any]:
    with pool.connection() as conn:
        order = _fetch_order(conn, order_id)
    if order is None:
        raise ApiError(404, "NOT_FOUND", "Order not found")
    return order


def create_order(pool: ConnectionPool, payload: dict[str, Any]) -> dict[str, Any]:
    with pool.connection() as conn:
        with conn.transaction():
            customer = conn.execute(queries.ACTIVE_CUSTOMER_EXISTS, (payload["customerId"],)).fetchone()
            if customer is None:
                raise ApiError(404, "NOT_FOUND", "Customer not found")

            address = conn.execute(
                queries.ADDRESS_BELONGS_TO_CUSTOMER,
                (payload["addressId"], payload["customerId"]),
            ).fetchone()
            if address is None:
                raise ApiError(404, "NOT_FOUND", "Address not found")

            order_id = conn.execute(
                queries.INSERT_ORDER,
                (payload["customerId"], payload["addressId"]),
            ).fetchone()["id"]

            for item in payload["items"]:
                product = conn.execute(queries.LOCK_PRODUCT, (item["productId"],)).fetchone()
                if product is None:
                    raise ApiError(404, "NOT_FOUND", "Product not found")
                if product["stock_quantity"] < item["quantity"]:
                    raise ApiError(409, "CONFLICT", "Insufficient stock")

                updated_stock = conn.execute(
                    queries.UPDATE_PRODUCT_STOCK,
                    (item["quantity"], item["productId"], item["quantity"]),
                ).fetchone()
                if updated_stock is None:
                    raise ApiError(409, "CONFLICT", "Insufficient stock")

                conn.execute(
                    queries.INSERT_ORDER_ITEM,
                    (order_id, item["productId"], item["quantity"], product["unit_price"]),
                )

            total = conn.execute(queries.UPDATE_ORDER_TOTAL, (order_id, order_id)).fetchone()["total_amount"]
            conn.execute(queries.INSERT_PAYMENT, (order_id, payload["payment"]["method"], total))
            conn.execute(
                queries.INSERT_AUDIT_LOG,
                ("order", order_id, "create_order", json.dumps(payload, ensure_ascii=False)),
            )

            order = _fetch_order(conn, order_id)

    if order is None:
        raise ApiError(500, "DATABASE_ERROR", "Order was created but could not be loaded")
    return order

File: apps/node-api/src/repository.js
import { ApiError } from "./errors.js";
import * as queries from "./queries.js";
import { customerFromRow, orderFromRows, productFromRow } from "./serializers.js";

async function fetchCustomer(client, customerId) {
  const result = await client.query(queries.GET_CUSTOMER, [customerId]);
  if (result.rowCount === 0) {
    return null;
  }
  return customerFromRow(result.rows[0]);
}

export async function getCustomer(pool, customerId) {
  const client = await pool.connect();
  try {
    const customer = await fetchCustomer(client, customerId);
    if (!customer) {
      throw new ApiError(404, "NOT_FOUND", "Customer not found");
    }
    return customer;
  } finally {
    client.release();
  }
}

export async function listCustomers(pool, page, pageSize) {
  const client = await pool.connect();
  try {
    const offset = (page - 1) * pageSize;
    const total = await client.query(queries.COUNT_CUSTOMERS);
    const rows = await client.query(queries.LIST_CUSTOMERS, [pageSize, offset]);
    return {
      page,
      pageSize,
      total: total.rows[0].total,
      items: rows.rows.map(customerFromRow)
    };
  } finally {
    client.release();
  }
}

export async function createCustomer(pool, payload) {
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    const inserted = await client.query(queries.INSERT_CUSTOMER, [
      payload.fullName,
      payload.email,
      payload.documentNumber,
      payload.phone
    ]);
    const customerId = inserted.rows[0].id;
    const address = payload.address;
    await client.query(queries.INSERT_ADDRESS, [
      customerId,
      address.label,
      address.street,
      address.number,
      address.complement,
      address.district,
      address.city,
      address.state,
      address.postalCode,
      address.isDefault
    ]);
    await client.query(queries.INSERT_AUDIT_LOG, [
      "customer",
      customerId,
      "create_customer",
      JSON.stringify(payload)
    ]);
    const customer = await fetchCustomer(client, customerId);
    await client.query("COMMIT");
    return customer;
  } catch (error) {
    await client.query("ROLLBACK");
    if (error.code === "23505") {
      throw new ApiError(409, "CONFLICT", "Customer email or document already exists");
    }
    throw error;
  } finally {
    client.release();
  }
}

export async function updateCustomer(pool, customerId, payload) {
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    const updated = await client.query(queries.UPDATE_CUSTOMER, [
      payload.fullName,
      payload.phone,
      payload.status,
      customerId
    ]);
    if (updated.rowCount === 0) {
      throw new ApiError(404, "NOT_FOUND", "Customer not found");
    }

    const address = payload.address;
    const changedAddress = await client.query(queries.UPDATE_DEFAULT_ADDRESS, [
      address.label,
      address.street,
      address.number,
      address.complement,
      address.district,
      address.city,
      address.state,
      address.postalCode,
      address.isDefault,
      customerId
    ]);
    if (changedAddress.rowCount === 0) {
      await client.query(queries.INSERT_ADDRESS, [
        customerId,
        address.label,
        address.street,
        address.number,
        address.complement,
        address.district,
        address.city,
        address.state,
        address.postalCode,
        address.isDefault
      ]);
    }

    await client.query(queries.INSERT_AUDIT_LOG, [
      "customer",
      customerId,
      "update_customer",
      JSON.stringify(payload)
    ]);
    const customer = await fetchCustomer(client, customerId);
    await client.query("COMMIT");
    return customer;
  } catch (error) {
    await client.query("ROLLBACK");
    throw error;
  } finally {
    client.release();
  }
}

export async function listProducts(pool, categoryId) {
  const client = await pool.connect();
  try {
    const category = await client.query(queries.CATEGORY_EXISTS, [categoryId]);
    if (category.rowCount === 0) {
      throw new ApiError(404, "NOT_FOUND", "Category not found");
    }
    const products = await client.query(queries.LIST_PRODUCTS_BY_CATEGORY, [categoryId]);
    return {
      categoryId,
      items: products.rows.map(productFromRow)
    };
  } finally {
    client.release();
  }
}

async function fetchOrder(client, orderId) {
  const result = await client.query(queries.GET_ORDER, [orderId]);
  return orderFromRows(result.rows);
}

export async function getOrder(pool, orderId) {
  const client = await pool.connect();
  try {
    const order = await fetchOrder(client, orderId);
    if (!order) {
      throw new ApiError(404, "NOT_FOUND", "Order not found");
    }
    return order;
  } finally {
    client.release();
  }
}

export async function createOrder(pool, payload) {
  const client = await pool.connect();
  try {
    await client.query("BEGIN");
    const customer = await client.query(queries.ACTIVE_CUSTOMER_EXISTS, [payload.customerId]);
    if (customer.rowCount === 0) {
      throw new ApiError(404, "NOT_FOUND", "Customer not found");
    }
    const address = await client.query(queries.ADDRESS_BELONGS_TO_CUSTOMER, [
      payload.addressId,
      payload.customerId
    ]);
    if (address.rowCount === 0) {
      throw new ApiError(404, "NOT_FOUND", "Address not found");
    }

    const inserted = await client.query(queries.INSERT_ORDER, [payload.customerId, payload.addressId]);
    const orderId = inserted.rows[0].id;

    for (const item of payload.items) {
      const product = await client.query(queries.LOCK_PRODUCT, [item.productId]);
      if (product.rowCount === 0) {
        throw new ApiError(404, "NOT_FOUND", "Product not found");
      }
      const productRow = product.rows[0];
      if (productRow.stock_quantity < item.quantity) {
        throw new ApiError(409, "CONFLICT", "Insufficient stock");
      }
      const stock = await client.query(queries.UPDATE_PRODUCT_STOCK, [
        item.quantity,
        item.productId,
        item.quantity
      ]);
      if (stock.rowCount === 0) {
        throw new ApiError(409, "CONFLICT", "Insufficient stock");
      }
      await client.query(queries.INSERT_ORDER_ITEM, [
        orderId,
        item.productId,
        item.quantity,
        productRow.unit_price
      ]);
    }

    const total = await client.query(queries.UPDATE_ORDER_TOTAL, [orderId, orderId]);
    await client.query(queries.INSERT_PAYMENT, [
      orderId,
      payload.payment.method,
      total.rows[0].total_amount
    ]);
    await client.query(queries.INSERT_AUDIT_LOG, [
      "order",
      orderId,
      "create_order",
      JSON.stringify(payload)
    ]);

    const order = await fetchOrder(client, orderId);
    await client.query("COMMIT");
    return order;
  } catch (error) {
    await client.query("ROLLBACK");
    throw error;
  } finally {
    client.release();
  }
}
